src/runtime/testing/sequence.ts

```diff
@@ -0,0 +1,594 @@
+ * keep state directly in your sequence test then it won't reset between orderings. You can:
+ * (a) register a per-ordering variable with sequenceTest.registerVariable()
+ * (b) set its value with sequenceTest.setVariable()
+ * (C) query the current value with sequenceTest.getVariable()
+ * 
+ * See storageNG/tests/store-sequence-test.ts for some fairly comprehensive examples of sequence testing
+ * in practise.
+ */
+export class SequenceTest<T> {
+  private prepareFunction: () => T;
+  private currentID = 0;
+
+  private changes: {[index: string]: SequenceChange[]} = {};
+
+  private inputs: Map<string, Input> = new Map();
+  private variables: Map<string, Variable> = new Map();
+  private sensors: Map<string, Sensor> = new Map();
+  private outputs: Map<string, Output> = new Map();
+
+  private interleavingLog: string[];
+
+  /**
+   * Set a function that constructs a fresh instance of the object under test for each ordering.
+   */
+  setTestConstructor(prepareFunction: () => T) {
+    this.prepareFunction = prepareFunction;
+  }
+
+  /**
+   * Register an input for the object under test. An input is a function that will be provided
+   * with a series of data updates.
+   * 
+   * @param name the function name to invoke in order to provide input
+   * @param asyncCount the number of internal awaits within the function
+   * @param response either {type: Void} (the function should not return a value),
+   *                 or {type: Constant, response: v} (the function should return v)
+   *                 or {type: Defer, checkResponse: (v) => void} (the return value of
+   *                         the function will be passed to checkResponse)
+   * 
+   * returns: an input key.
+   */
+  registerInput(name: string, asyncCount: number, response: InputResponseObject): string {
+    const key = name + this.currentID++;
+    this.inputs.set(key, {name, delay: asyncCount, response, results: []});
+    return key;
+  }
+
+  /**
+   * Register an output for the object under test. An output is a function that will be
+   * invoked as the object is fed changes. The parameters provided to the output can be
+   * inspected as part of the test.
+   * 
+   * @param name the function name to replace or register with
+   * @param response either {type: Void} (the function won't return a value),
+   *                 or {type: Constant, response: v} (the function will return v),
+   *                 or {type: Defer, default: v} (the function will return v initially).
+   *                 The response object can also contain an onOutput: (v) => void which
+   *                 can be used to check the value provided to the output function.
+   * @param behavior either Register (the function of name `name` will be invoked in order
+   *                                  to register an output function)
+   *                 or Replace (the function of name `name` will be replaced with an output
+   *                             function)
```

Comment:
This is a bit hard to understand. From the code in `setupOutputs`, I gather Replace means a method on the object under test will be swapped with something generated by the test framework, which will then verify the calls to that method (i.e. by the object under test) according to the response options provided above.

I'm still not sure exactly how Register works.

Reply:
Added some examples to clarify.